<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>TODO for splicejam • splicejam</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script>




<meta property="og:title" content="TODO for splicejam" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-title-body">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.69.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
    <li>
      <a href="articles/shiny-server-for-sashimi-plots.html">Shiny server for Sashimi Plots</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/jmw86069/splicejam/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>TODO for splicejam</h1>
    </div>

<div id="todo-for-splicejam" class="section level1">

<div id="16jul2021" class="section level2">
<h2 class="hasAnchor">
<a href="#16jul2021" class="anchor"></a>16jul2021</h2>
<ul>
<li>
<p>COMPLETE: Update (minor): Import STAR <code>"SJ.out.tab"</code> format equivalent to importing BED12 format.</p>
<ul>
<li>When the junctions file has 9 columns it is assumed to be STAR format.</li>
<li>The import now uses <code><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">data.table::fread()</a></code> and not <code><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html">rtracklayer::import.bed()</a></code>.</li>
</ul>
</li>
<li>
<p>COMPLETE: Update (minor): Optional R-shiny startup without displaying a gene.</p>
<ul>
<li>Use <code>default_gene="blank"</code> to start with a blank plot.</li>
</ul>
</li>
<li>
<p>Test (minor): test against alternative, commonly used GTF sources</p>
<ul>
<li>Use case: Most steps were designed to use Gencode GTF. The workflow should also work when using other GTF files.</li>
<li>Notable areas to test: Many steps assume GTF annotations include “gene_name”, “transcript_id” - but if not present the user should be able to define alternate names.</li>
</ul>
</li>
</ul>
</div>
<div id="13jul2021" class="section level2">
<h2 class="hasAnchor">
<a href="#13jul2021" class="anchor"></a>13jul2021</h2>
<ul>
<li>
<p><code>makeTx2geneDFfromGtf()</code> extend to include genome coordinates (major)</p>
<ul>
<li>currently <code>tx2geneDF</code> does not return genome coordinates, in part because a given gene symbol may be located in multiple locations, multiple chromosomes, e.g. <code>"7SK"</code>.</li>
</ul>
</li>
<li>
<p>Consider bookdown online documentation (major)</p>
<ul>
<li>Need model to follow for where to put documentation, how to host. Check: jokergoo/ComplexHeatmap; clusterProfiler, enrichplot.</li>
</ul>
</li>
<li>
<p>Consider using patchwork instead of cowplot?</p>
<ul>
<li>There is some issue with including the gene model panel twice, maybe patchwork does that step well?</li>
</ul>
</li>
</ul>
</div>
<div id="12jul2021" class="section level2">
<h2 class="hasAnchor">
<a href="#12jul2021" class="anchor"></a>12jul2021</h2>
<ul>
<li>
<p>Update (minor): Update Shiny progress bar after coverage is complete, before splice junctions are being loaded.</p>
<ul>
<li>Use case: After loading coverage files, for example the label says “coverage (16 of 16)”, it pauses for a long time. What is it doing? I want to know!</li>
</ul>
</li>
<li>
<p>New feature (minor): Allow resizing the gene-transcript panel</p>
<ul>
<li>Use case: Sometimes there are way too many transcripts, the panel should be much taller. Sometimes the exon labels are clipped off the bottom of the figure.</li>
<li>Note that adjusting “per panel size” affects gene-transcript panel, making it much too crowded.</li>
</ul>
</li>
<li>
<p>New feature (minor): Allow down-sampling the coverage profile</p>
<ul>
<li>Use case: exons are rarely displayed one base per pixel, more often an exon is roughly 100-500 bases wide, but displayed in the equivalent of 25-50 pixels. The adjustment would be to down-sample data to help speed up the rendering step.</li>
</ul>
</li>
<li>
<p>New feature (minor): Optionally display gene model per column for multi-column layout.</p>
<ul>
<li>Use case: Currently with two-column display the gene model does not visually align to the coverages.</li>
</ul>
</li>
<li>
<p>New feature (inquiry): Test porting to ggbio or Gviz for broader re-use.</p>
<ul>
<li>Use case: To integrate other track types, coverage, peaks, genes, etc. the ggbio and Gviz R packages are more feature-rich. I think ggbio is no longer officially supported?</li>
<li>General idea is to provide two functionalities: track “geom”, and x-axis compressed axis which effectively compresses the intron ranges.</li>
<li>For Gviz, understand <code>GdObject-class</code> and create CustomTrack, which should mimic the workflow used by <code>AlignmentsTrack()</code>:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu">Gviz</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/CustomTrack-class.html">CustomTrack</a></span><span class="op">(</span>plottingFunction<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">GdObject</span>, <span class="va">prepare</span><span class="op">=</span><span class="cn">FALSE</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span><span class="op">}</span>,
   variables<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>, name<span class="op">=</span><span class="st">"CustomTrack"</span>, <span class="va">...</span><span class="op">)</span></pre></div>
</li>
<li>
<p>New feature (minor): Color picker alongside Shiny app “Sample Selection”</p>
<ul>
<li>Use case is to display, and allow selection of colors per sample</li>
<li>Not sure if “easy” color picker widgets will work with the selectable table widget.</li>
</ul>
</li>
<li>
<p>New feature (major): allow multiple genes/features in display region, not just one gene</p>
<ul>
<li>Use case: display more than one gene; display genes and peaks.</li>
<li>Use case: allow zooming out - beyond the annotated gene.</li>
<li>Use case: display other genes that overlap the display coordinates.</li>
<li>Changes:</li>
<li>Define the display region by genome coordinates, not by the gene symbol - which itself only implies coordinates.</li>
<li>Define compressed ranges using the features displayed. This step also requires renaming the compressed ranges something more broadly useful. If exons from two genes overlap, what shall we name this region? Probably should use “region_1” or “gap_1”.</li>
<li>Option to compress ranges using only certain features - outside the Shiny app, this step is already possible using <code><a href="reference/make_ref2compressed.html">splicejam::make_ref2compressed()</a></code> with a set of <code>GRanges</code>.</li>
<li>Major: junction strandedness should not be forced to match the strand of the displayed gene. This change may cause problems using STAR junctions, since STAR junction strandedness is not always accurate.</li>
</ul>
</li>
<li>
<p>New feature (moderate): Extend <code><a href="reference/make_ref2compressed.html">splicejam::make_ref2compressed()</a></code></p>
<ul>
<li>Use case: If displaying RNA-seq and ChIP-seq data together, it may be useful to extend regions around the TSS before compressing gaps, to show signal near the TSS. Similar with exons, peaks, etc. Some configuration options could be useful to describe specifically in this function.</li>
</ul>
</li>
<li>
<p>New feature (major): Display multiple genes in adjacent panels.</p>
<ul>
<li>Use case: Given 3 genes of interest, display sashimi plots side-by-side</li>
<li>Counterpoint: This step can be done “manually” by preparing individual sashimi plots, then using <code>cowplot</code> or <code>patchwork</code> to assemble them into multi-panel figure.</li>
<li>Conclusion: Probably will not implement. May need to add this use case to the vignette.</li>
</ul>
</li>
<li>
<p>New feature (major): display Sashimi plots similar to ggridges ridge plots</p>
<ul>
<li>Use case: ggplot2 facet panels take up much extra space, making figures look heavy and bulky - they take up a lot of web browser space. The idea is to plot coverage/junctions slightly overlapping so they are adjacent and perhaps more easily compared visually. Also, more samples can be plotted together without using so much figure space.</li>
<li>Changes:</li>
<li>The y-axis label is only practical for the first sample.</li>
<li>Draw baseline for each sample, render back samples first, top-to-bottom.</li>
<li>Define flexible offset, either by fraction of max height, or by absolute y-axis coverage units.</li>
<li>Allow optional ggplot2 facets, which could still be used to split different experiment design groups. Or could split by sample and display each sample replicate per panel.</li>
<li>Could allow some trickery, like ordering samples by a factor column, where visual gaps could be inserted by using empty factor levels. For example: GroupA_Veh, GroupA_Treated, GroupA_blank, GroupB_Veh, GroupB_Treated. In this case <code>"GroupA_blank"</code> would be a factor level, but there would be no sashimi plot data for that factor level, so it would be drawn empty, leaving a visual gap.</li>
</ul>
</li>
<li>
<p>New features (major): coverage and junctions from BAM files</p>
<ul>
<li>Use case is to allow BAM input, both for coverage and for junctions.</li>
<li>Major enabling feature: This step is a precursor to handling scRNA-seq, which could dynamically split the BAM reads by tSNE/UMAP clusters.</li>
</ul>
</li>
<li>
<p>New data (major): pre-defined hg19 transcriptome data required for sashimi plots</p>
<ul>
<li>Gencode comprehensive with all genes, flat exons pre-computed.</li>
<li>Obvious enhancement: hg38, mm10, other common organisms.</li>
<li>Describe steps used so others can use their own GTF as needed.</li>
</ul>
</li>
<li>
<p>New feature (nice to have): ability to select/deselect transcripts displayed in the R-shiny app for a given gene.</p>
<p>Use case is when displaying all transcripts for a given gene, but there are too many un-expressed transcripts, e.g. human ACTB. Could allow a separate table to select transcripts to include/hide. When performing “Update” the compressed exons would be re-calculated for that gene using the selected exons.</p>
</li>
</ul>
</div>
<div id="19mar2021-todo-items" class="section level2">
<h2 class="hasAnchor">
<a href="#19mar2021-todo-items" class="anchor"></a>19mar2021 todo items</h2>
<ul>
<li>
<p>COMPLETE: Fix broken Sample Selection - change to table selection/ordering</p>
<ul>
<li>include <code><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo()</a></code> in a hidden section on the R-shiny app to be able to confirm all packages and version numbers being used.</li>
</ul>
</li>
</ul>
</div>
<div id="18mar2021-todo-items" class="section level2">
<h2 class="hasAnchor">
<a href="#18mar2021-todo-items" class="anchor"></a>18mar2021 todo items</h2>
<ul>
<li><p>COMPLETE: Fix the ggrepel exon labeling which is now above instead of below the gene-exon model.</p></li>
<li>
<p>for interactive plots, highlighting a splice junction also highlights all splice junctions for the same exon span, which is good</p>
<ul>
<li>for flat gene-exon models, name the gap using the format <code>geneSymbolexon_name1-geneSymbolexon_name2</code> which will allow the gap to be highlighted as well.</li>
<li>for flat transcript-exon models, not sure if this is appropriate.</li>
</ul>
</li>
</ul>
</div>
<div id="23apr2020-todo-items" class="section level2">
<h2 class="hasAnchor">
<a href="#23apr2020-todo-items" class="anchor"></a>23apr2020 todo items</h2>
<ul>
<li><p>Consider adding transcript selection to the left menu items</p></li>
<li>
<p>Workflow:</p>
<ul>
<li>Idea is to allow manually selecting subset or superset of transcripts to include in the flatExonsByGene, partly to allow removing transcripts that mess up the overall gene model, partly to allow showing a highlighted subset of transcripts for example for diffSplice/DEXseq splice hits. Sometimes a gene has 8 detected isoforms, but the predominant change involves only 2 or 3 of those isoforms. Would be great to be able to create the ideal subset.</li>
<li>(Note this workflow is already possible with manual steps outside the R-shiny app, but who wants to do that?)</li>
</ul>
</li>
<li><p>Allow expanding the x-axis genomic region being displayed, beyond the gene body. Note this step requires defining the compression to use for upstream region – for example should it be 10:1 compressed in visual space?</p></li>
<li>
<p>Note the genomic coordinates are not being displayed below certain genes (human GAS5) – why not? It might be because this gene has one contiguous exon due to some noisy unspliced isoforms. The coordinate labeling function might be trying to use only the outside coordinate. Could adjust that function to use either the edge, or if two labels are separated by more then 5% the total exon width, display that label.</p>
<ul>
<li>Genomic axis label logic for compressed GRanges coordinates: Start by labeling outer edge of each exon, then internal exon boundaries. Add each label if it will be at least 5% distant from another label, based upon the total exon width, and total number of gaps.</li>
<li>Review new ggplot2 axis labeling rules and whether we can rely upon ggplot2 to hide axis labels that would otherwise overlap.</li>
</ul>
</li>
<li><p>Allow manually setting one common y-axis range. Currently only possible to adjust the y-axis by using interactive plotly, which loses some important junction labeling.</p></li>
<li><p>Allow displaying any gene-exon model that sits inside the viewing range, in case there is an interesting internal feature such as a miRNA. This enhancement is the predecessor for supplying a genomic region and having the sashimi plot generated in that region regardless of genes involved. Note that junction strandedness will need to be more fluid – currently the junction strand is fixed consistent with the target gene, mostly because STAR junction strandedness is sometimes mis-annotated (i.e. 100% coverage on “+” strand, but one junction might be labeled “-” strand.)</p></li>
<li><p>Pie in the sky idea: some way to save per-gene settings, like coordinate range, axis ranges, etc. So when someone goes to “Gria1” they get the best default Gria1 view they can get. When they go to “Smarca4” they get the best default Smarca4 view they can get.</p></li>
</ul>
</div>
<div id="21apr2020-todo-items" class="section level2">
<h2 class="hasAnchor">
<a href="#21apr2020-todo-items" class="anchor"></a>21apr2020 todo items</h2>
<ul>
<li><p>COMPLETE: It looks like junction stacking on negative strand is applied to the wrong edge of each junction (left-right instead of right-left).</p></li>
<li>
<p>COMPLETE: Allow the gene-transcript panel to be adjusted taller.</p>
<ul>
<li>By default, gene panel height should be auto-adjusted based upon the number of transcripts shown.</li>
<li>Ideally the panel height should be adjusted based upon label height.</li>
</ul>
</li>
<li>
<p>Enable more default settings such as number of columns, panel height, font sizes, etc. The app should be fully customizable upon startup, so that new users get the intended default experience.</p>
<ul>
<li>
<p>Available:</p>
<ul>
<li>min_junction_reads - minimum junction read threshold</li>
<li>use_exon_names - one of <code><a href="https://rdrr.io/r/base/c.html">c("coordinates", "exon names")</a></code>
</li>
<li>exon_range_selected - two gene_nameExon values</li>
<li>exon_range_choices_default - exon labels available for default gene</li>
<li>layout_ncol - number of columns in layout</li>
<li>include_strand - one of <code><a href="https://rdrr.io/r/base/c.html">c("+","-","both")</a></code> strands to display</li>
<li>share_y_axis - one of <code><a href="https://rdrr.io/r/base/c.html">c(TRUE, FALSE)</a></code> whether to use shared y-axis range</li>
</ul>
</li>
<li>
<p>Todo:</p>
<ul>
<li>panel height</li>
<li>show gene model</li>
<li>show transcripts</li>
<li>detected only</li>
<li>font size,</li>
<li>exon font size</li>
<li>interactive plot</li>
<li>minimum arc height? new option</li>
</ul>
</li>
</ul>
</li>
<li><p>Consider minimum junction threshold based upon % coverage range instead of absolute number. Not urgent, can be manually adjusted.</p></li>
<li><p>Consider option to label gene exons by major exon number (exon1, exon2, etc.) not using the disjoint exon numbers (exon1a, exon1b, exon2a, exon2b, etc.)</p></li>
<li>
<p><code><a href="reference/detectedTxInfo.html">detectedTxInfo()</a></code></p>
<ul>
<li>option to return only detected results.</li>
<li>change labels to be user-friendly</li>
</ul>
</li>
<li>
<p>new function to convert <code><a href="reference/detectedTxInfo.html">detectedTxInfo()</a></code> into ComplexHeatmap:</p>
<ul>
<li>box around cells which meet the “detected” thresholds overall (cell_fun)</li>
<li>color cells by percent max expression per sample (default), option to use count, TPM</li>
<li>label cells: pseudo-counts (percent max expr, TPM abundance); 67.9 (13%, 1.2 TPM)</li>
<li>option to display “detected” or “all” Tx</li>
</ul>
</li>
<li><p>Consider optional per-panel label to supplement/replace the text label used by ggplot2 facet wrap. Label could be positioned topright, topleft, bottomright, bottomleft, or auto.</p></li>
<li><p>Two-column displays are not visually synchronized with the gene view, consider making a two-column gene view?</p></li>
<li><p>When using plotly interactive plots with two columns, consider syncing the two columns when zooming in one side, so the other column is also zoomed into the same region.</p></li>
<li><p>Consider adding two-column output selection to the main page.</p></li>
<li><p>When viewing very large genes in plotly interactive mode, it can take ages to render (after the ggplot object is already created, it appears the slow step is creating all the plotly-specific javascript to render every coordinate position, even though the region is zoomed out.) Needs some optimization step to reduce the number of actual x-axis points being displayed. SMARCA4 is an example gene, spans 150k bases, yet every exon pixel is stored as a coordinate for plotly. Consider reducing the resolution of x-axis points being rendered by plotly somehow.</p></li>
</ul>
</div>
<div id="bug-fixes" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes" class="anchor"></a>Bug fixes</h2>
<ul>
<li>
<code><a href="reference/prepareSashimi.html">prepareSashimi()</a></code> throws an error when <code>flatExonsByGene</code> contains different seqlevels than the <code>BigWigFile</code>, usually when <code>flatExonsByGene</code> contains more seqlevels than present in the <code>BigWigFile</code>. It happens even when the exon features of interest involve seqlevels present in <code>BigWigFile</code>. The fix is to reduce the seqlevels in <code>flatExonsByGene</code> to match those in the <code>BigWigFile</code>, or to reduce seqlevels to actual features in <code>flatExonsByGene</code>. At that point an error would accurately reflect that <code>BigWigFile</code> does not contain seqlevels for the requested seqlevels, and thus would be helpful.</li>
</ul>
<div id="sashimi-usability-enhancements" class="section level3">
<h3 class="hasAnchor">
<a href="#sashimi-usability-enhancements" class="anchor"></a>Sashimi usability enhancements</h3>
<ul>
<li>Separate the function <code><a href="reference/sashimiAppConstants.html">sashimiAppConstants()</a></code> to its own proper standalone function, that prepares all the dependency data objects like <code>flatExonsByGene</code>, <code>flatExonsByTx</code>, <code>tx2geneDF</code>.</li>
<li>New function to assemble <code>filesDF</code>. Essentially appends a file to an existing <code>data.frame</code>.</li>
<li>New function <code>plotlySashimi()</code> that optionally includes the gene-transcript-exon model in the visualization.</li>
<li>Zoom by exon range, using exon names. For example, <code>zoom_by=c("Gria1_exon11", "Gria1_exon16")</code>.</li>
</ul>
</div>
<div id="sashimi-longer-term-enhancements" class="section level3">
<h3 class="hasAnchor">
<a href="#sashimi-longer-term-enhancements" class="anchor"></a>Sashimi longer term enhancements</h3>
<ul>
<li>Enable display of sample replicates. Actually, enable the grouping of <code>sample_id</code> entries. Default is to overlay replicates in the same facet panel, facet by <code>sample_group</code>. Alternative is to offset the y-axis similar to the <code>"ggridges"</code> package. Unclear if the layers coverage polygons, and junction arcs, will be out of sync; that is sample_1 should draw coverage and junctions, before sample_2 is drawn.</li>
<li>Enable BAM file for coverage data. Defer filtering rules to the <code>Rsamtools</code> package, but show recommended examples for properly paired reads.</li>
<li>Enable BAM file for junction data. Determine recommended filters for properly paired reads with minimum overlap width at each end of the junction.</li>
<li>Enable Sashimi plots over a region, not just one gene range. Refactor of <code><a href="reference/prepareSashimi.html">prepareSashimi()</a></code> to take GRanges range an optional coordinate range, or multiple genes (which would imply the coordinate range if on the same chromosome).</li>
<li>Create an example showing two genes with overlapping exons.</li>
<li>Mock up an example that combines gene exons, ChIP-seq peaks, and ATAC-seq peaks, together into one schematic of a genome region. It should show before and after, compressing the “intervening sequences” between features.</li>
<li>Mock up example showing <code><a href="reference/detectedTxInfo.html">detectedTxInfo()</a></code> output in the form of plots. For example, show percent max expression as a heatmap, each call labeled to show the counts, TPM, percentage, and whether it was called “detected” by the thresholds.</li>
</ul>
</div>
<div id="farris-r-shiny-feedback" class="section level3">
<h3 class="hasAnchor">
<a href="#farris-r-shiny-feedback" class="anchor"></a>Farris R-shiny feedback</h3>
<ul>
<li>Recalculate scale factors, use reference genes from in situ hybridizations: Calm1, Camk2a, Camk2b, Actb, which show little to no change.</li>
<li>Plotly two-column layout does not keep all columns of panels in sync when the x-axis is zoomed, the desired effect is to have the x-axis of each sashimi panel zoomed consistently, across however many columns are used for layout.</li>
<li>Gria1 takes a while to display the first time, during which the app is not usable. Consider making plot rendering asynchronous to decouple the visualization from the use of the R-shiny app.</li>
<li>The plotly exon coordinates show the overall coordinate range and not the specific exon.</li>
</ul>
</div>
<div id="vignettes" class="section level3">
<h3 class="hasAnchor">
<a href="#vignettes" class="anchor"></a>Vignettes</h3>
<ul>
<li>New vignette describing how to start a new R-shiny Sashimi App.</li>
</ul>
</div>
<div id="testing" class="section level3">
<h3 class="hasAnchor">
<a href="#testing" class="anchor"></a>Testing</h3>
<ul>
<li>Implement methods in the testthat package to ensure a consistent suite of tests to confirm full functionality with updates. Highest priority functions, which would break parts of the workflow:
<ul>
<li><code><a href="reference/flattenExonsBy.html">flattenExonsBy()</a></code></li>
<li>
<code><a href="reference/assignGRLexonNames.html">assignGRLexonNames()</a></code> - dependent upon <code><a href="https://rdrr.io/pkg/jamba/man/makeNames.html">jamba::makeNames()</a></code>
</li>
<li>
<code><a href="reference/annotateGRfromGR.html">annotateGRfromGR()</a></code> - dependent upon <code><a href="reference/shrinkMatrix.html">shrinkMatrix()</a></code>
</li>
<li><code><a href="reference/shrinkMatrix.html">shrinkMatrix()</a></code></li>
<li><code><a href="reference/combineGRcoverage.html">combineGRcoverage()</a></code></li>
<li><code><a href="reference/df2colorSub.html">df2colorSub()</a></code></li>
</ul>
</li>
</ul>
</div>
<div id="dataframe-versus-tibble" class="section level3">
<h3 class="hasAnchor">
<a href="#dataframe-versus-tibble" class="anchor"></a>data.frame versus tibble</h3>
<ul>
<li>Evaluated tibble as possible replacement for data.frame in <code><a href="reference/prepareSashimi.html">prepareSashimi()</a></code>. It shrinks data volume to about 7 times smaller, reducing duplicated annotations per row for polygon coordinates. However, ggplot2 still requires unnesting the tibble into a tall format for plotting, it is unclear whether that step will incur its own performance hit.</li>
<li>ggplot2 polygon rendering should be faster – possibly needs to reduce the polygon detail prior to plotting, but unclear whether that can happen since ggplot2 already chops polygons into multiple small segments in order to handle transformed axis coordinates.</li>
</ul>
</div>
<div id="r-shiny-caching" class="section level3">
<h3 class="hasAnchor">
<a href="#r-shiny-caching" class="anchor"></a>R-shiny caching</h3>
<ul>
<li>Review flatExonsByTx caching, which gets invalidated too frequently. Consider using a fixed caching strategy, using defined keys instead of the memoise default.</li>
</ul>
</div>
<div id="r-shiny-launchsashimiapp" class="section level3">
<h3 class="hasAnchor">
<a href="#r-shiny-launchsashimiapp" class="anchor"></a>R-shiny launchSashimiApp()</h3>
<ul>
<li><p>Better error message for gene symbol not recognized.</p></li>
<li>
<p>Plotly tooltips are so finicky, but try anyway:</p>
<ul>
<li>coverage tooltip: y-position (coverage score), exon name (gr_name), track name (cov_name), preferably the coordinate range of the feature.</li>
<li>junction tooltip: junction score, name “gene_exonFrom gene_exonTo”, and start-end coordinates of the junction.</li>
<li>plotly_onclick: show junction score or exon name in the middle of the respective polygon?</li>
</ul>
</li>
<li><p>Make font sizes more configurable.</p></li>
<li><p>Add gene symbol somewhere, otherwise without the gene model shown, there is no indication what gene is being displayed.</p></li>
<li><p>Add a vignette describing how to set up an R-shiny splicejam.</p></li>
</ul>
</div>
<div id="sashimi-plot-functionality" class="section level3">
<h3 class="hasAnchor">
<a href="#sashimi-plot-functionality" class="anchor"></a>Sashimi plot functionality</h3>
<ul>
<li><p>refactor prepareSashimi() to create full merged data.frame, thus removing it from plotSashimi(); it should make plot faster.</p></li>
<li><p>geom_diagonal_wide_arc(): new argument calc_midpoint=TRUE will calculate the appropriate middle x-position using the active x-axis transformation so the mid-point is the visible middle. Optionally specify the fraction, 0.5 is middle, 0.25 is the first quartile near the left edge, etc.</p></li>
<li>
<p>Transformed axis:</p>
<ul>
<li>customize breaks, minor_breaks, and labels functions so ggplot2 labels include as many non-overlapping x-axis labels at exon boundaries as possible. Make minor_breaks include tick marks and optional smaller labels when zoomed in, or when an exon is wide enough to have an intermediate label.</li>
<li>Transformed axis line: x-axis line indicating compresssed status, solid line = uncompressed, dotted line = compressed. It might need a new own geom.</li>
</ul>
</li>
</ul>
</div>
<div id="portabilityusability" class="section level3">
<h3 class="hasAnchor">
<a href="#portabilityusability" class="anchor"></a>Portability/Usability</h3>
<ul>
<li>Verify “Plan B for Windows users.” Test and confirm whether using Windows Subsystem for Linux (WSL) will enable bigWig import.</li>
<li>Expand input types to allow BAM files, from which coverage will be calculated. Will need an appropriate caching strategy, probably at the level of covGR GRanges with RleList to store coverage in value columns.</li>
<li>Expand junction input to allow BAM files, from which junctions can be inferred. (Check existing Bioconductor packages for any that convert BAM to junctions.)</li>
</ul>
</div>
<div id="vignettes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#vignettes-1" class="anchor"></a>Vignettes</h3>
<ul>
<li>
<p>More Sashimi examples:</p>
<ul>
<li>Using baseline to “shift” a set of exons up on the y-axis, use Ntrk2 as an example.</li>
<li>Use baseline with Gria1 to shift up/down the two alternate exons. (The original vision was something like a Sankey plot.)</li>
</ul>
</li>
<li><p>ALE-specific RNA-seq analysis workflow.</p></li>
<li><p>Codon usage analysis workflow.</p></li>
</ul>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward, Shannon Farris.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


