<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Shiny server for Sashimi Plots • splicejam</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Shiny server for Sashimi Plots">
<meta property="og:description" content="splicejam">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.69.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
    <li>
      <a href="../articles/shiny-server-for-sashimi-plots.html">Shiny server for Sashimi Plots</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/jmw86069/splicejam/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="shiny-server-for-sashimi-plots_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Shiny server for Sashimi Plots</h1>
                        <h4 class="author">James M. Ward</h4>
            
            <h4 class="date">2021-07-27</h4>
      
      <small class="dont-index">Source: <a href="http://github.com/jmw86069/splicejam/blob/master/vignettes/shiny-server-for-sashimi-plots.Rmd"><code>vignettes/shiny-server-for-sashimi-plots.Rmd</code></a></small>
      <div class="hidden name"><code>shiny-server-for-sashimi-plots.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>This vignette is intended to describe how to create a Shiny server to display interactive Sashimi plots with RNA-seq data.</p>
<p>Splicejam uses memoise to cache data, so it is recommended to start the Splicejam Shiny app in its own directory. Each type of data is cached in its own sub-directory with “_memoise” in the name, so you can recognize and delete these directories to clear the cache as needed.</p>
</div>
<div id="data-requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#data-requirements" class="anchor"></a>Data Requirements</h2>
<p>There are two basic requirements for a Sashimi plot:</p>
<ol style="list-style-type: decimal">
<li><p>Gene-exon structure, usually provided by a GTF file.</p></li>
<li>
<p>Source data, sequence coverage and junction read counts:</p>
<ul>
<li>RNA-seq coverage data in bigWig files.</li>
<li>Splice junction data in BED files or STAR <code>"SJ.out.tab"</code> files.</li>
</ul>
</li>
</ol>
<p>Briefly:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(splicejam)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># define files with coverage and junctions</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>filesDF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample_id=</span><span class="fu">c</span>(<span class="st">"sample_A"</span>, <span class="st">"sample_A"</span>),</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">url=</span><span class="fu">c</span>(<span class="st">"https://server/sample_A.bw"</span>, <span class="st">"https://server/sample_A/SJ.out.tab"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">type=</span><span class="fu">c</span>(<span class="st">"bw"</span>, <span class="st">"junction"</span>));</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># provide path to genes GTF</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> <span class="st">"path/to/genes.gtf"</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># launch Splicejam Shiny app</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">launchSashimiApp</span>()</span></code></pre></div>
</div>
<div id="gene-exon-structure-using-gtf-file" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-exon-structure-using-gtf-file" class="anchor"></a>Gene-exon structure using GTF file</h2>
<p>Ideally, the GTF used by splicejam will be the same GTF file used in upstream processing, for example with Salmon, Kallisto, or featureCounts. The benefit is that the GTF will display gene-exon structure consistent with your overall analysis work.</p>
<p>That said, any GTF file for your genome will work fine. The GTF is used to determine gene-exon structure, and to display transcript isoforms per gene. It then displays RNA-seq coverage and junction reads over these exons.</p>
<p>Splicejam will derive several objects from this GTF file:</p>
<ul>
<li>
<strong>tx2geneDF</strong> - <code>data.frame</code> with transcript-to-gene association. Specifically it looks for <code>"gene_name"</code> and <code>"transcript_id"</code>.</li>
<li>
<strong>flatExonsByGene</strong> - exons flattened for each gene. If optional <code>detectedTx</code> is provided, then it will include only exons from the detected transcripts.</li>
</ul>
<p>Most of the workflow was designed for the Gencode GTF format, which includes <code>"gene_name"</code> with gene symbols, and <code>"transcript_id"</code> with transcript identifiers.</p>
<p>Other GTF files should work even if they have custom gene and transcript attributes. When in doubt, use a Gencode GTF file.</p>
<div id="commonly-used-gtf-files-by-organism" class="section level3">
<h3 class="hasAnchor">
<a href="#commonly-used-gtf-files-by-organism" class="anchor"></a>Commonly used GTF files by organism</h3>
<p>Some GTF files that have been tested and confirmed with Splicejam are listed below.</p>
<pre><code>## Mouse mm10 as used by Farris et al
ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M12/gencode.vM12.annotation.gtf.gz

## Mouse mm10
http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/gencode.vM27.annotation.gtf.gz

## Human hg38
http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/gencode.v38.annotation.gtf.gz

## Human hg19
http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/GRCh37_mapping/gencode.v38lift37.annotation.gtf.gz</code></pre>
</div>
<div id="example-using-gencode-for-hg19" class="section level3">
<h3 class="hasAnchor">
<a href="#example-using-gencode-for-hg19" class="anchor"></a>Example using Gencode for hg19</h3>
<p>Simple enough, just assign the file path to <code>gtf</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">gtf</span> <span class="op">&lt;-</span> <span class="st">"http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/GRCh37_mapping/gencode.v38lift37.annotation.gtf.gz"</span>;</pre></div>
</div>
</div>
<div id="source-data" class="section level2">
<h2 class="hasAnchor">
<a href="#source-data" class="anchor"></a>Source data</h2>
<p>Source data with sequence coverage and junction read counts, is supplied as a <code>data.frame</code> with these columns:</p>
<ul>
<li>
<code>"url"</code>: The web URL or file path to each file.</li>
<li>
<code>"sample_id"</code>: The name of each sample as it should appear in each panel.</li>
<li>
<code>"type"</code>: A <code>character</code> value with either <code>"bw"</code> for bigWig coverage, or <code>"junction"</code> for junction read counts.</li>
<li>
<code>"scale_factor"</code>: optional <code>numeric</code> column, used to scale numeric values for each files. This column is used for dynamic normalization, for example if you have size factors from DESeq2, they can be used here. The default <code>scale_factor=1</code>.</li>
<li>Other columns may be included.</li>
</ul>
<p>When there are multiple replicate files for a <code>sample_id</code>, the scores are added together and the total score is displayed in the sashimi plot for each <code>sample_id</code>. When <code>scale_factor</code> is also defined, it is applied to each file before values are summed across <code>sample_id</code> replicates. In this way, files can be individually normalized as needed.</p>
<div id="example-filesdf-data-frame" class="section level3">
<h3 class="hasAnchor">
<a href="#example-filesdf-data-frame" class="anchor"></a>Example filesDF data.frame</h3>
<p>An example of <code>filesDF</code> is provided in the R package <code>"farrisdata"</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">if</span> <span class="op">(</span><span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jamba/man/check_pkg_installed.html">check_pkg_installed</a></span><span class="op">(</span><span class="st">"farrisdata"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
   <span class="fu">farrisdata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/farrisdata/man/farris_sashimi_files_df.html">farris_sashimi_files_df</a></span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="op">}</span>
<span class="co">#&gt;    sample_id</span>
<span class="co">#&gt; 1     CA1_CB</span>
<span class="co">#&gt; 2     CA1_DE</span>
<span class="co">#&gt; 3     CA2_CB</span>
<span class="co">#&gt; 4     CA2_DE</span>
<span class="co">#&gt; 5     CA3_CB</span>
<span class="co">#&gt; 6     CA3_DE</span>
<span class="co">#&gt; 7      DG_CB</span>
<span class="co">#&gt; 8      DG_DE</span>
<span class="co">#&gt; 21    CA1_CB</span>
<span class="co">#&gt; 17    CA1_CB</span>
<span class="co">#&gt; 41    CA1_DE</span>
<span class="co">#&gt; 31    CA1_DE</span>
<span class="co">#&gt; 61    CA2_CB</span>
<span class="co">#&gt; 51    CA2_CB</span>
<span class="co">#&gt; 81    CA2_DE</span>
<span class="co">#&gt; 71    CA2_DE</span>
<span class="co">#&gt; 10    CA3_CB</span>
<span class="co">#&gt; 9     CA3_CB</span>
<span class="co">#&gt; 12    CA3_DE</span>
<span class="co">#&gt; 11    CA3_DE</span>
<span class="co">#&gt; 14     DG_CB</span>
<span class="co">#&gt; 13     DG_CB</span>
<span class="co">#&gt; 16     DG_DE</span>
<span class="co">#&gt; 15     DG_DE</span>
<span class="co">#&gt;                                                                                             url</span>
<span class="co">#&gt; 1  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA1_CB.STAR_mm10.combinedJunctions.bed.gz</span>
<span class="co">#&gt; 2  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA1_DE.STAR_mm10.combinedJunctions.bed.gz</span>
<span class="co">#&gt; 3  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA2_CB.STAR_mm10.combinedJunctions.bed.gz</span>
<span class="co">#&gt; 4  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA2_DE.STAR_mm10.combinedJunctions.bed.gz</span>
<span class="co">#&gt; 5  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA3_CB.STAR_mm10.combinedJunctions.bed.gz</span>
<span class="co">#&gt; 6  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA3_DE.STAR_mm10.combinedJunctions.bed.gz</span>
<span class="co">#&gt; 7   https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/DG_CB.STAR_mm10.combinedJunctions.bed.gz</span>
<span class="co">#&gt; 8   https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/DG_DE.STAR_mm10.combinedJunctions.bed.gz</span>
<span class="co">#&gt; 21          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA1_CB.union.neg.bw</span>
<span class="co">#&gt; 17          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA1_CB.union.pos.bw</span>
<span class="co">#&gt; 41          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA1_DE.union.neg.bw</span>
<span class="co">#&gt; 31          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA1_DE.union.pos.bw</span>
<span class="co">#&gt; 61          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA2_CB.union.neg.bw</span>
<span class="co">#&gt; 51          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA2_CB.union.pos.bw</span>
<span class="co">#&gt; 81          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA2_DE.union.neg.bw</span>
<span class="co">#&gt; 71          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA2_DE.union.pos.bw</span>
<span class="co">#&gt; 10          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA3_CB.union.neg.bw</span>
<span class="co">#&gt; 9           https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA3_CB.union.pos.bw</span>
<span class="co">#&gt; 12          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA3_DE.union.neg.bw</span>
<span class="co">#&gt; 11          https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/CA3_DE.union.pos.bw</span>
<span class="co">#&gt; 14           https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/DG_CB.union.neg.bw</span>
<span class="co">#&gt; 13           https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/DG_CB.union.pos.bw</span>
<span class="co">#&gt; 16           https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/DG_DE.union.neg.bw</span>
<span class="co">#&gt; 15           https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/union_bigwig/DG_DE.union.pos.bw</span>
<span class="co">#&gt;        type scale_factor</span>
<span class="co">#&gt; 1  junction    0.8796491</span>
<span class="co">#&gt; 2  junction    1.1064974</span>
<span class="co">#&gt; 3  junction    0.8461586</span>
<span class="co">#&gt; 4  junction    1.6993700</span>
<span class="co">#&gt; 5  junction    0.8615462</span>
<span class="co">#&gt; 6  junction    1.1941632</span>
<span class="co">#&gt; 7  junction    0.7705711</span>
<span class="co">#&gt; 8  junction    1.2941457</span>
<span class="co">#&gt; 21       bw    0.9891766</span>
<span class="co">#&gt; 17       bw    0.9891766</span>
<span class="co">#&gt; 41       bw    1.2362753</span>
<span class="co">#&gt; 31       bw    1.2362753</span>
<span class="co">#&gt; 61       bw    0.9817312</span>
<span class="co">#&gt; 51       bw    0.9817312</span>
<span class="co">#&gt; 81       bw    1.0678433</span>
<span class="co">#&gt; 71       bw    1.0678433</span>
<span class="co">#&gt; 10       bw    1.0058324</span>
<span class="co">#&gt; 9        bw    1.0058324</span>
<span class="co">#&gt; 12       bw    1.4419289</span>
<span class="co">#&gt; 11       bw    1.4419289</span>
<span class="co">#&gt; 14       bw    0.8449864</span>
<span class="co">#&gt; 13       bw    0.8449864</span>
<span class="co">#&gt; 16       bw    0.8218802</span>
<span class="co">#&gt; 15       bw    0.8218802</span></pre></div>
</div>
<div id="rna-seq-coverage-data" class="section level3">
<h3 class="hasAnchor">
<a href="#rna-seq-coverage-data" class="anchor"></a>RNA-seq coverage data</h3>
<p>RNA-seq coverage is provided as bigWig files, and can be accessed via HTTP web hyperlink, or a direct file path.</p>
</div>
<div id="splice-junction-data" class="section level3">
<h3 class="hasAnchor">
<a href="#splice-junction-data" class="anchor"></a>Splice junction data</h3>
<p>Splice junctions are provided in one of two formats:</p>
<ol style="list-style-type: decimal">
<li>
<strong>BED format</strong>: We use BED12 format, but it can be direct BED format as well. The BED12 format is used with two 1-base alignments and a gap between them to indicate the splice junction itself. When all BED names are numeric, the BED name is used as the junction score, otherwise the BED score is used. The reason is that bigBed format does not allow scores higher than 1000, so we encode scores in the name field.</li>
<li>
<strong>STAR <code>"SJ.out.tab"</code> format</strong>: a tab-delimited file produced by the STAR alignment tool. This file must have 9 columns, and column 7 is used to define read counts because it contains uniquely mapped reads. Junctions with zero uniquely mapped reads are removed.</li>
</ol>
</div>
</div>
<div id="quick-start-using-farris-data" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start-using-farris-data" class="anchor"></a>Quick start using Farris data</h2>
<p>As a positive control for the Splicejam Shiny server, you can use the Farris data that supports <em>Farris et al (2019)</em>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jamba/man/check_pkg_installed.html">check_pkg_installed</a></span><span class="op">(</span><span class="st">"farrisdata"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
   <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"jmw86069/farrisdata"</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/jmw86069/splicejam">splicejam</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/launchSashimiApp.html">launchSashimiApp</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p>Note this workflow will use <code>filesDF</code> from the “farrisdata” package, and will download Gencode mouse GTF used for that publication. It takes about 3 minutes to prepare data and create the first Shiny plot for the gene <strong>Gria1</strong>.</p>
<p>This workflow will by default populate the R environment <code><a href="https://rdrr.io/r/base/environment.html">globalenv()</a></code> with variables used in the farrisdata Splicejam Shiny app. See Advanced Options for ways to specify a new environment.</p>
</div>
<div id="quick-start-with-gtf-and-filesdf" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start-with-gtf-and-filesdf" class="anchor"></a>Quick start with GTF and filesDF</h2>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/jmw86069/splicejam">splicejam</a></span><span class="op">)</span>

<span class="co"># filesDF should already be defined</span>
<span class="va">filesDF2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="fu">farrisdata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/farrisdata/man/farris_sashimi_files_df.html">farris_sashimi_files_df</a></span>,
   <span class="va">sample_id</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CA1_DE"</span>, <span class="st">"CA2_CB"</span><span class="op">)</span><span class="op">)</span>;

<span class="co"># gtf should be a file path or web URL</span>
<span class="va">gtf</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M12/gencode.vM12.annotation.gtf.gz"</span>;

<span class="fu"><a href="../reference/launchSashimiApp.html">launchSashimiApp</a></span><span class="op">(</span>filesDF<span class="op">=</span><span class="va">filesDF2</span>, gtf<span class="op">=</span><span class="va">gtf</span><span class="op">)</span></pre></div>
<p>This workflow will by default populate the R environment <code><a href="https://rdrr.io/r/base/environment.html">globalenv()</a></code> with variables used in the farrisdata Splicejam Shiny app. See Advanced Options for ways to specify a new environment.</p>
</div>
<div id="useful-custom-options" class="section level2">
<h2 class="hasAnchor">
<a href="#useful-custom-options" class="anchor"></a>Useful custom options</h2>
<p>The following options can be invoked by defining the variable name inside the enrivonment used for the Splicejam Shiny app. By default, this environment is <code><a href="https://rdrr.io/r/base/environment.html">globalenv()</a></code>, however the examples below show how to use a custom environment. The advantage of using an environment is that the data contained inside is not copied in memory during function calls, and can be shared by the Shiny UI and Shiny Server.</p>
<ul>
<li>
<strong>detectedTx</strong> - a <code>character</code> vector of <code>transcript_id</code> values that were “detected” by your experiment. You can define these however you prefer, or you can leave this value blank and include every transcript in the GTF file. Supplying a subset of detectedTx is beneficial by presenting a simpler gene-exon structure per gene. It also speeds up the initial Splicejam Shiny server start up time.</li>
<li>
<strong>detectedGenes</strong> - a <code>character</code> vector of <code>gene_name</code> values that were “detected” by your experiment. The main effect of <code>detectedGenes</code> is that it limits the number of flat gene-exons prepared by Splicejam Shiny, and limits the number of genes that can be searched. However, the user can still change <strong>“Genes to Search”</strong> to <strong>“All genes”</strong> to search all genes defined by the GTF file. In that case, each new gene will have exons flattened during load time. This process usually takes only about 1 second longer than normal per gene, in return for having a faster Splicejam Shiny server start up time the first time.</li>
<li>
<strong>color_sub</strong> - a <code>character</code> vector of colors, with names that match your <code>sample_id</code> values. This vector will define colors used in the Sashimi plots.</li>
</ul>
<p>For example, to define <code>detectedTx</code> you would simply assign a value in the <code><a href="https://rdrr.io/r/base/environment.html">globalenv()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">detectedTx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tx2geneDF</span><span class="op">)</span>;</pre></div>
<p>Or if using a custom environment:</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">splicejam_env</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html">new.env</a></span><span class="op">(</span><span class="op">)</span>;
<span class="fu"><a href="https://rdrr.io/r/base/assign.html">assign</a></span><span class="op">(</span><span class="st">"detectedTx"</span>,
   envir<span class="op">=</span><span class="va">splicejam_env</span>,
   value<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tx2geneDF</span><span class="op">)</span><span class="op">)</span>;</pre></div>
</div>
<div id="advanced-options" class="section level2">
<h2 class="hasAnchor">
<a href="#advanced-options" class="anchor"></a>Advanced options</h2>
<div id="start-the-shiny-app-on-a-different-port" class="section level3">
<h3 class="hasAnchor">
<a href="#start-the-shiny-app-on-a-different-port" class="anchor"></a>Start the Shiny app on a different port</h3>
<p>One of the most common ways to set up a Shiny server is to run it on a custom port, and listen to a specific address.</p>
<p>Note in the example below, <code>host="0.0.0.0"</code> will instruct the Shiny app to respond to requests directed at any host or IP address. If you used <code>host="127.0.0.1"</code> the server would only respond to requests specific to <code>https://127.0.0.1:8080</code> and would not respond to requests to <code>https://localhost:8080</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="../reference/launchSashimiApp.html">launchSashimiApp</a></span><span class="op">(</span>options<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
   port<span class="op">=</span><span class="fl">8080</span>,
   hist<span class="op">=</span><span class="st">"0.0.0.0"</span><span class="op">)</span><span class="op">)</span></pre></div>
</div>
<div id="specify-a-specific-r-environment-for-splicejam-data" class="section level3">
<h3 class="hasAnchor">
<a href="#specify-a-specific-r-environment-for-splicejam-data" class="anchor"></a>Specify a specific R environment for Splicejam data</h3>
<p>By default the data preparation uses the global environment defined by <code><a href="https://rdrr.io/r/base/environment.html">globalenv()</a></code>. This process will create objects in your user session, and will update those objects during the preparation step.</p>
<p>However, you can create a custom environment to keep the data encapsulated, and separate from your user session.</p>
<p>It is intended to be straightforward to use a custom environment. First define a new environment.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/jmw86069/splicejam">splicejam</a></span><span class="op">)</span>

<span class="va">splicejam_env</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html">new.env</a></span><span class="op">(</span><span class="op">)</span>;

<span class="co"># filesDF should already be defined</span>
<span class="va">filesDF2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="fu">farrisdata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/farrisdata/man/farris_sashimi_files_df.html">farris_sashimi_files_df</a></span>,
   <span class="va">sample_id</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CA1_DE"</span>, <span class="st">"CA2_CB"</span><span class="op">)</span><span class="op">)</span>;

<span class="co"># gtf should be a file path or web URL</span>
<span class="va">gtf</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M12/gencode.vM12.annotation.gtf.gz"</span>;

<span class="fu"><a href="../reference/launchSashimiApp.html">launchSashimiApp</a></span><span class="op">(</span>
   empty_uses_farrisdata<span class="op">=</span><span class="cn">FALSE</span>,
   envir<span class="op">=</span><span class="va">splicejam_env</span>,
   filesDF<span class="op">=</span><span class="va">filesDF2</span>,
   gtf<span class="op">=</span><span class="va">gtf</span><span class="op">)</span></pre></div>
</div>
<div id="specify-gene-exon-data-without-using-a-gtf-file" class="section level3">
<h3 class="hasAnchor">
<a href="#specify-gene-exon-data-without-using-a-gtf-file" class="anchor"></a>Specify gene-exon data without using a GTF file</h3>
<p>At its core, splicejam derives the data it needs from the GTF data, but it can be supplied with this data directly, avoiding the need for a GTF file.</p>
<p>More detail will be added here in future, but for now see the vignette <strong>“Create a Sashimi Plot”</strong>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by James M. Ward, Shannon Farris.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
