<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Annotate GRanges using another GRanges object — annotateGRfromGR • splicejam</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Annotate GRanges using another GRanges object — annotateGRfromGR" />

<meta property="og:description" content="Annotate GRanges using another GRanges object" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.48.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Annotate GRanges using another GRanges object</h1>
    <small class="dont-index">Source: <a href='https://github.com/jmw86069/splicejam/blob/master/R/jambio-tx.R'><code>R/jambio-tx.R</code></a></small>
    <div class="hidden name"><code>annotateGRfromGR.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Annotate GRanges using another GRanges object</p>
    
    </div>

    <pre class="usage"><span class='fu'>annotateGRfromGR</span>(<span class='no'>GR1</span>, <span class='no'>GR2</span>, <span class='kw'>grOL</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>numAsStrings</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>stringShrinkFunc</span> <span class='kw'>=</span> <span class='kw'>function</span>(<span class='no'>...</span>) {     <span class='kw pkg'>jamba</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/jamba/topics/cPasteUnique'>cPasteUnique</a></span>(<span class='no'>...</span>, <span class='kw'>doSort</span>
  <span class='kw'>=</span> <span class='fl'>TRUE</span>) }, <span class='kw'>numShrinkFunc</span> <span class='kw'>=</span> <span class='no'>sum</span>, <span class='kw'>addStringCols</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>type</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"any"</span>, <span class='st'>"start"</span>, <span class='st'>"end"</span>, <span class='st'>"within"</span>, <span class='st'>"equal"</span>),
  <span class='kw'>ignore.strand</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>select</span> <span class='kw'>=</span> <span class='st'>"all"</span>, <span class='kw'>sep</span> <span class='kw'>=</span> <span class='st'>","</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>DEBUG</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>GR1</th>
      <td><p>GRanges object, the reference object to which annotations
are added.</p></td>
    </tr>
    <tr>
      <th>GR2</th>
      <td><p>GRanges object used for annotations</p></td>
    </tr>
    <tr>
      <th>grOL</th>
      <td><p>overlaps object, optionally used when the
<code><a href='https://www.rdocumentation.org/packages/GenomicRanges/topics/findOverlaps-methods'>GenomicRanges::findOverlaps()</a></code> function has already been run, mostly
intended to save re-processing the overlaps for large objects.</p></td>
    </tr>
    <tr>
      <th>numAsStrings</th>
      <td><p>logical indicating whether numerical values should
be treated as strings for the purpose of added annotations. When
<code>TRUE</code>, numerical values will be converted to character strings and
retained as if they were labels. This argument treats all numeric
columns as string, to treat only a subset use the <code>addStringCols</code>
argument.</p></td>
    </tr>
    <tr>
      <th>stringShrinkFunc</th>
      <td><p>function used to shrink string annotations
by overlapping feature. This function should take a list as input,
and <code>sep</code> as an argument indicating the delimiter if applicable,
and return a character vector of the same length as the list. By
default, <code><a href='https://www.rdocumentation.org/packages/jamba/topics/cPasteUnique'>jamba::cPasteUnique()</a></code> is used. Control over the sort
and uniqueness should be applied with this function.
Note: <code>stringShrinkFunc</code> can be a single function, or can be a
named list of function calls, whose names match the colnames
of <code>values</code>. If a named list is provided, the first entry is used
for any any names in <code>values</code> which are not in <code><a href='https://www.rdocumentation.org/packages/base/topics/names'>names(stringShrinkFunc)</a></code>.</p></td>
    </tr>
    <tr>
      <th>numShrinkFunc</th>
      <td><p>function used to shrink numerical values
by overlapping feature. For numeric values, the <code>data.table</code> package
is used to apply the function, which enables custom functions not
otherwise available via the <code>GenomicRanges</code> methods. Therefore, the
function does not take a list as input, instead takes a numeric
vector as input.
Note: <code>numShrinkFunc</code> can be a single function, or can be a
named list of function calls, whose names match the colnames
of <code>values</code>. If a named list is provided, the first entry is used
for any any names in <code>values</code> which are not in <code><a href='https://www.rdocumentation.org/packages/base/topics/names'>names(numShrinkFunc)</a></code>.</p></td>
    </tr>
    <tr>
      <th>addStringCols</th>
      <td><p>character vector, optional, of numeric colnames that
should be treated as string values. This option is a subset of the
<code>numAsStrings</code>.</p></td>
    </tr>
    <tr>
      <th>type, ignore.strand, select</th>
      <td><p>arguments sent to
<code><a href='https://www.rdocumentation.org/packages/GenomicRanges/topics/findOverlaps-methods'>GenomicRanges::findOverlaps()</a></code>. Note these options are only used
when <code>grOL</code> is not supplied.</p></td>
    </tr>
    <tr>
      <th>sep</th>
      <td><p>character value indicating the delimiter to use between
string values.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical indicating whether to print verbose output.</p></td>
    </tr>
    <tr>
      <th>DEBUG</th>
      <td><p>logical indicating whether to print debugging output.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments are ignored.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>GRanges object with colnames added to <code>values</code>, with length
and order equal to the input <code>GR1</code> GRanges object.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function adds annotations to features in the given GRanges
object, from overlapping features in a second GRanges object.
It is especially useful after performing a manipulation that
results in a GRanges object without any <code>values</code> annotation,
for example <code><a href='https://www.rdocumentation.org/packages/GenomicRanges/topics/inter-range-methods'>GenomicRanges::reduce()</a></code> or <code><a href='https://www.rdocumentation.org/packages/GenomicRanges/topics/setops-methods'>GenomicRanges::intersect()</a></code>.</p>
<p>In theory this function is relatively simple, it applies annotations
to overlapping entries. In practice, it gets complicated when multiple
annotations overlap the same GRange entry. In that case, numerical
values by default return the <code><a href='https://www.rdocumentation.org/packages/base/topics/mean'>base::mean()</a></code> while string values call
<code><a href='https://www.rdocumentation.org/packages/jamba/topics/cPasteUnique'>jamba::cPasteUnique()</a></code> by default, and return the unique, sorted,
comma-delimited set of string values. For example, overlapping several
exons from the same gene, the resulting annotation might include
just one gene symbol.</p>
<p>The numeric values can be aggregated using another function, controlled
by <code>numShrinkFunction</code> named using the colname. For example:
<code>numShrinkFunction="min"</code> would return the <code><a href='https://www.rdocumentation.org/packages/base/topics/Extremes'>base::min()</a></code> value for all
numeric columns. But <code>numShrinkFunction=list(default=mean, score=max)</code>
would return the <code><a href='https://www.rdocumentation.org/packages/base/topics/mean'>base::mean()</a></code> for all numeric columns except the
<code>"score"</code> column which would report the <code><a href='https://www.rdocumentation.org/packages/base/topics/Extremes'>base::max()</a></code>.</p>
<p>Numeric values currently use the <code>data.table</code> package workflow, which
provides extremely efficient calculations for numeric values. However,
vectorized functions have the potential to be notably faster, as is
true with <code><a href='https://www.rdocumentation.org/packages/jamba/topics/cPasteUnique'>jamba::cPasteUnique()</a></code> especially when applying
<code><a href='https://www.rdocumentation.org/packages/jamba/topics/mixedSort'>jamba::mixedSort()</a></code> to sort alphanumeric values. In future, the
implementation may change to accomodate vectorized numeric functions,
instead of using <code>data.table</code>.</p>
<p>TODO: This function is a specific case where another function
<code>shrinkDataFrame()</code> may be a more general purpose use. That function
is yet to be ported to the Jam package suite.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other jam GRanges functions: <code><a href='addGRLgaps.html'>addGRLgaps</a></code>,
  <code><a href='addGRgaps.html'>addGRgaps</a></code>,
  <code><a href='annotateGRLfromGRL.html'>annotateGRLfromGRL</a></code>,
  <code><a href='assignGRLexonNames.html'>assignGRLexonNames</a></code>,
  <code><a href='closestExonToJunctions.html'>closestExonToJunctions</a></code>,
  <code><a href='combineGRcoverage.html'>combineGRcoverage</a></code>,
  <code><a href='exoncov2polygon.html'>exoncov2polygon</a></code>,
  <code><a href='findOverlapsGRL.html'>findOverlapsGRL</a></code>,
  <code><a href='getFirstStrandedFromGRL.html'>getFirstStrandedFromGRL</a></code>,
  <code><a href='getGRLgaps.html'>getGRLgaps</a></code>,
  <code><a href='getGRcoverageFromBw.html'>getGRcoverageFromBw</a></code>,
  <code><a href='getGRgaps.html'>getGRgaps</a></code>, <code><a href='grl2df.html'>grl2df</a></code>,
  <code><a href='make_ref2compressed.html'>make_ref2compressed</a></code>, <code><a href='sortGRL.html'>sortGRL</a></code>,
  <code><a href='spliceGR2junctionDF.html'>spliceGR2junctionDF</a></code>,
  <code><a href='stackJunctions.html'>stackJunctions</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>gr12</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/GenomicRanges/topics/GRanges-class'>GRanges</a></span>(
   <span class='kw'>seqnames</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"chr1"</span>, <span class='st'>"chr2"</span>, <span class='st'>"chr1"</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>3</span>,<span class='fl'>3</span>,<span class='fl'>3</span>)),
   <span class='kw'>ranges</span><span class='kw'>=</span><span class='kw pkg'>IRanges</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/IRanges/topics/IRanges-constructor'>IRanges</a></span>(
      <span class='kw'>start</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>100</span>, <span class='fl'>200</span>, <span class='fl'>400</span>, <span class='fl'>500</span>, <span class='fl'>300</span>, <span class='fl'>100</span>, <span class='fl'>200</span>, <span class='fl'>400</span>, <span class='fl'>600</span>),
      <span class='kw'>width</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>100</span>,<span class='fl'>150</span>,<span class='fl'>50</span>, <span class='fl'>50</span>,<span class='fl'>50</span>,<span class='fl'>100</span>, <span class='fl'>50</span>,<span class='fl'>200</span>,<span class='fl'>50</span>)
   ),
   <span class='kw'>strand</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"+"</span>, <span class='st'>"-"</span>, <span class='st'>"+"</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>3</span>,<span class='fl'>3</span>,<span class='fl'>3</span>)),
   <span class='kw'>gene_name</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"GeneA"</span>, <span class='st'>"GeneB"</span>, <span class='st'>"GeneC"</span>), <span class='kw'>each</span><span class='kw'>=</span><span class='fl'>3</span>)
)
<span class='no'>gr1</span> <span class='kw'>&lt;-</span> <span class='no'>gr12</span>[,<span class='fl'>0</span>];

<span class='co'># Say for example you have a GRanges object with no annotation</span>
<span class='no'>gr1</span>;</div><div class='output co'>#&gt; GRanges object with 9 ranges and 0 metadata columns:
#&gt;       seqnames    ranges strand
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
#&gt;   [1]     chr1   100-199      +
#&gt;   [2]     chr1   200-349      +
#&gt;   [3]     chr1   400-449      +
#&gt;   [4]     chr2   500-549      -
#&gt;   [5]     chr2   300-349      -
#&gt;   [6]     chr2   100-199      -
#&gt;   [7]     chr1   200-249      +
#&gt;   [8]     chr1   400-599      +
#&gt;   [9]     chr1   600-649      +
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
<span class='co'># And had another GRanges object with annotations</span>
<span class='no'>gr12</span>;</div><div class='output co'>#&gt; GRanges object with 9 ranges and 1 metadata column:
#&gt;       seqnames    ranges strand |   gene_name
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#&gt;   [1]     chr1   100-199      + |       GeneA
#&gt;   [2]     chr1   200-349      + |       GeneA
#&gt;   [3]     chr1   400-449      + |       GeneA
#&gt;   [4]     chr2   500-549      - |       GeneB
#&gt;   [5]     chr2   300-349      - |       GeneB
#&gt;   [6]     chr2   100-199      - |       GeneB
#&gt;   [7]     chr1   200-249      + |       GeneC
#&gt;   [8]     chr1   400-599      + |       GeneC
#&gt;   [9]     chr1   600-649      + |       GeneC
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
<span class='co'># To add annotations</span>
<span class='fu'>annotateGRfromGR</span>(<span class='no'>gr1</span>, <span class='no'>gr12</span>);</div><div class='output co'>#&gt; GRanges object with 9 ranges and 1 metadata column:
#&gt;         seqnames    ranges strand |   gene_name
#&gt;            &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#&gt;   GR1_1     chr1   100-199      + |       GeneA
#&gt;   GR1_2     chr1   200-349      + | GeneA,GeneC
#&gt;   GR1_3     chr1   400-449      + | GeneA,GeneC
#&gt;   GR1_4     chr2   500-549      - |       GeneB
#&gt;   GR1_5     chr2   300-349      - |       GeneB
#&gt;   GR1_6     chr2   100-199      - |       GeneB
#&gt;   GR1_7     chr1   200-249      + | GeneA,GeneC
#&gt;   GR1_8     chr1   400-599      + | GeneA,GeneC
#&gt;   GR1_9     chr1   600-649      + |       GeneC
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
<span class='co'># Notice certain features overlap multiple annotations,</span>
<span class='co'># which may be acceptable.</span>

<span class='co'># If you want to keep annotations distinct,</span>
<span class='co'># use annotateGRLfromGRL()</span>
<span class='no'>grl1</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'>split</span>(<span class='no'>gr12</span>[,<span class='fl'>0</span>],
   <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'>values</span>(<span class='no'>gr12</span>)$<span class='no'>gene_name</span>);
<span class='no'>grl2</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'>split</span>(<span class='no'>gr12</span>,
   <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'>values</span>(<span class='no'>gr12</span>)$<span class='no'>gene_name</span>);

<span class='co'># The first object is a GRangesList with no annotations</span>
<span class='no'>grl1</span>;</div><div class='output co'>#&gt; GRangesList object of length 3:
#&gt; $GeneA 
#&gt; GRanges object with 3 ranges and 0 metadata columns:
#&gt;       seqnames    ranges strand
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
#&gt;   [1]     chr1   100-199      +
#&gt;   [2]     chr1   200-349      +
#&gt;   [3]     chr1   400-449      +
#&gt; 
#&gt; $GeneB 
#&gt; GRanges object with 3 ranges and 0 metadata columns:
#&gt;       seqnames  ranges strand
#&gt;   [1]     chr2 500-549      -
#&gt;   [2]     chr2 300-349      -
#&gt;   [3]     chr2 100-199      -
#&gt; 
#&gt; $GeneC 
#&gt; GRanges object with 3 ranges and 0 metadata columns:
#&gt;       seqnames  ranges strand
#&gt;   [1]     chr1 200-249      +
#&gt;   [2]     chr1 400-599      +
#&gt;   [3]     chr1 600-649      +
#&gt; 
#&gt; -------
#&gt; seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
<span class='co'># The second object is a GRangesList with annotation,</span>
<span class='co'># assumed to be in the same order</span>
<span class='no'>grl2</span>;</div><div class='output co'>#&gt; GRangesList object of length 3:
#&gt; $GeneA 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;       seqnames    ranges strand |   gene_name
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#&gt;   [1]     chr1   100-199      + |       GeneA
#&gt;   [2]     chr1   200-349      + |       GeneA
#&gt;   [3]     chr1   400-449      + |       GeneA
#&gt; 
#&gt; $GeneB 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;       seqnames  ranges strand | gene_name
#&gt;   [1]     chr2 500-549      - |     GeneB
#&gt;   [2]     chr2 300-349      - |     GeneB
#&gt;   [3]     chr2 100-199      - |     GeneB
#&gt; 
#&gt; $GeneC 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;       seqnames  ranges strand | gene_name
#&gt;   [1]     chr1 200-249      + |     GeneC
#&gt;   [2]     chr1 400-599      + |     GeneC
#&gt;   [3]     chr1 600-649      + |     GeneC
#&gt; 
#&gt; -------
#&gt; seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
<span class='fu'><a href='annotateGRLfromGRL.html'>annotateGRLfromGRL</a></span>(<span class='no'>grl1</span>, <span class='no'>grl2</span>);</div><div class='output co'>#&gt; GRangesList object of length 3:
#&gt; $GeneA 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;           seqnames    ranges strand |           X
#&gt;              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#&gt;   grl1_v1     chr1   100-199      + |       GeneA
#&gt;   grl1_v2     chr1   200-349      + |       GeneA
#&gt;   grl1_v3     chr1   400-449      + |       GeneA
#&gt; 
#&gt; $GeneB 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;           seqnames  ranges strand |     X
#&gt;   grl1_v4     chr2 500-549      - | GeneB
#&gt;   grl1_v5     chr2 300-349      - | GeneB
#&gt;   grl1_v6     chr2 100-199      - | GeneB
#&gt; 
#&gt; $GeneC 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;           seqnames  ranges strand |     X
#&gt;   grl1_v7     chr1 200-249      + | GeneC
#&gt;   grl1_v8     chr1 400-599      + | GeneC
#&gt;   grl1_v9     chr1 600-649      + | GeneC
#&gt; 
#&gt; -------
#&gt; seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

