<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ splicejam</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.69.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
    <li>
      <a href="../articles/shiny-server-for-sashimi-plots.html">Shiny server for Sashimi Plots</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/jmw86069/splicejam/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-splicejam-core-functions" class="hasAnchor"><a href="#section-splicejam-core-functions" class="anchor"></a>Splicejam core functions</h2>
          <p class="section-desc"><p>These functions are the heart of the splicejam package</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="exoncov2polygon.html">exoncov2polygon()</a></code> </p>
        </td>
        <td><p>Convert exon coverage to polygons</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene2gg.html">gene2gg()</a></code> </p>
        </td>
        <td><p>Gene GRangesList to ggplot2 grob</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="grl2df.html">grl2df()</a></code> </p>
        </td>
        <td><p>GRangesList to data.frame for ggplot2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_ref2compressed.html">make_ref2compressed()</a></code> </p>
        </td>
        <td><p>Create a ref2compressed function to compress GR gaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSashimi.html">plotSashimi()</a></code> </p>
        </td>
        <td><p>Jam Sashimi plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareSashimi.html">prepareSashimi()</a></code> </p>
        </td>
        <td><p>Prepare Sashimi plot data</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-rna-seq-functions" class="hasAnchor"><a href="#section-rna-seq-functions" class="anchor"></a>RNA-seq functions</h2>
          <p class="section-desc"><p>Functions aimed at RNA-seq analysis</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="assignGRLexonNames.html">assignGRLexonNames()</a></code> </p>
        </td>
        <td><p>Assign exon names to GRangesList</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="closestExonToJunctions.html">closestExonToJunctions()</a></code> </p>
        </td>
        <td><p>Find closest exon to splice junction ends</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combineGRcoverage.html">combineGRcoverage()</a></code> </p>
        </td>
        <td><p>Combine GRanges coverage replicates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="defineDetectedTx.html">defineDetectedTx()</a></code> </p>
        </td>
        <td><p>Define detected transcripts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectedTxInfo.html">detectedTxInfo()</a></code> </p>
        </td>
        <td><p>Summarize detected transcript results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="exoncov2polygon.html">exoncov2polygon()</a></code> </p>
        </td>
        <td><p>Convert exon coverage to polygons</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flattenExonsBy.html">flattenExonsBy()</a></code> </p>
        </td>
        <td><p>Flatten exons by gene or transcript</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGRcoverageFromBw.html">getGRcoverageFromBw()</a></code> </p>
        </td>
        <td><p>Get coverage for GRanges from bigWig files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="groups2contrasts.html">groups2contrasts()</a></code> </p>
        </td>
        <td><p>Define experimental contrasts from sample groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="internal_junc_score.html">internal_junc_score()</a></code> </p>
        </td>
        <td><p>maximum overlapping internal junction score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeTx2geneFromGtf.html">makeTx2geneFromGtf()</a></code> </p>
        </td>
        <td><p>Make tx2gene data.frame from a GTF file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_ref2compressed.html">make_ref2compressed()</a></code> </p>
        </td>
        <td><p>Create a ref2compressed function to compress GR gaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareSashimi.html">prepareSashimi()</a></code> </p>
        </td>
        <td><p>Prepare Sashimi plot data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runDiffSplice.html">runDiffSplice()</a></code> </p>
        </td>
        <td><p>Perform differential isoform analysis using diffSplice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sortSamples.html">sortSamples()</a></code> </p>
        </td>
        <td><p>Sort biological sample labels for experimental design</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spliceGR2junctionDF.html">spliceGR2junctionDF()</a></code> </p>
        </td>
        <td><p>Splice junction data.frame summary</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-splicejam-r-shiny-functions" class="hasAnchor"><a href="#section-splicejam-r-shiny-functions" class="anchor"></a>Splicejam R-shiny functions</h2>
          <p class="section-desc"><p>Functions to enable Splicejam R-shiny</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="launchSashimiApp.html">launchSashimiApp()</a></code> </p>
        </td>
        <td><p>Launch Sashimi R-shiny application</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sashimiAppConstants.html">sashimiAppConstants()</a></code> </p>
        </td>
        <td><p>Sashimi Shiny app constants</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sashimiAppServer.html">sashimiAppServer()</a></code> </p>
        </td>
        <td><p>Sashimi Shiny app server</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sashimiAppUI.html">sashimiAppUI()</a></code> </p>
        </td>
        <td><p>Sashimi Shiny app UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sashimiDataConstants.html">sashimiDataConstants()</a></code> </p>
        </td>
        <td><p>Prepare sashimi plot required data</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-ale-specific-rna-seq-functions" class="hasAnchor"><a href="#section-ale-specific-rna-seq-functions" class="anchor"></a>ALE-specific RNA-seq functions</h2>
          <p class="section-desc"><p>Functions specific to ALE (alternate last exon) RNA-seq analysis</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="ale2violin.html">ale2violin()</a></code> </p>
        </td>
        <td><p>Prepare ALE data for violin plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getFirstStrandedFromGRL.html">getFirstStrandedFromGRL()</a></code> </p>
        </td>
        <td><p>Get first stranded GRanges feature per GRangesList</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tx2ale.html">tx2ale()</a></code> </p>
        </td>
        <td><p>tx2ale: detect alternative last exons (ALE) from transcript data</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-rna-seq-codon-usage-functions" class="hasAnchor"><a href="#section-rna-seq-codon-usage-functions" class="anchor"></a>RNA-seq codon usage functions</h2>
          <p class="section-desc"><p>Functions specific to codon usage in RNA-seq analysis</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="codonUsage2df.html">codonUsage2df()</a></code> </p>
        </td>
        <td><p>Make codon usage data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dna2codon.html">dna2codon()</a></code> </p>
        </td>
        <td><p>Convert DNA to 3-base codons</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="jamCai.html">jamCai()</a></code> </p>
        </td>
        <td><p>Calculate Codon Adaptation Index</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-granges-functions" class="hasAnchor"><a href="#section-granges-functions" class="anchor"></a>GRanges functions</h2>
          <p class="section-desc"><p>Functions to manipulate the Genomic Ranges (GRanges) family of objects</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="addGRLgaps.html">addGRLgaps()</a></code> </p>
        </td>
        <td><p>Add gaps between GRangesList regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addGRgaps.html">addGRgaps()</a></code> </p>
        </td>
        <td><p>Add gaps between GRanges regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotateGRLfromGRL.html">annotateGRLfromGRL()</a></code> </p>
        </td>
        <td><p>Annotate GRangesList from GRangesList objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotateGRfromGR.html">annotateGRfromGR()</a></code> </p>
        </td>
        <td><p>Annotate GRanges using another GRanges object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assignGRLexonNames.html">assignGRLexonNames()</a></code> </p>
        </td>
        <td><p>Assign exon names to GRangesList</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="closestExonToJunctions.html">closestExonToJunctions()</a></code> </p>
        </td>
        <td><p>Find closest exon to splice junction ends</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combineGRcoverage.html">combineGRcoverage()</a></code> </p>
        </td>
        <td><p>Combine GRanges coverage replicates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="exoncov2polygon.html">exoncov2polygon()</a></code> </p>
        </td>
        <td><p>Convert exon coverage to polygons</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findOverlapsGRL.html">findOverlapsGRL()</a></code> </p>
        </td>
        <td><p>Find overlaps between two GRangesList objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flattenExonsBy.html">flattenExonsBy()</a></code> </p>
        </td>
        <td><p>Flatten exons by gene or transcript</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getFirstStrandedFromGRL.html">getFirstStrandedFromGRL()</a></code> </p>
        </td>
        <td><p>Get first stranded GRanges feature per GRangesList</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGRLgaps.html">getGRLgaps()</a></code> </p>
        </td>
        <td><p>Get gaps in GRangesList objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGRcoverageFromBw.html">getGRcoverageFromBw()</a></code> </p>
        </td>
        <td><p>Get coverage for GRanges from bigWig files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGRgaps.html">getGRgaps()</a></code> </p>
        </td>
        <td><p>Get gaps in GRanges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="grl2df.html">grl2df()</a></code> </p>
        </td>
        <td><p>GRangesList to data.frame for ggplot2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="jam_isDisjoint.html">jam_isDisjoint()</a></code> </p>
        </td>
        <td><p>Test whether GRanges are disjoint (non-overlapping)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_ref2compressed.html">make_ref2compressed()</a></code> </p>
        </td>
        <td><p>Create a ref2compressed function to compress GR gaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sortGRL.html">sortGRL()</a></code> </p>
        </td>
        <td><p>Sort GRangesList elements by chromosome and position</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spliceGR2junctionDF.html">spliceGR2junctionDF()</a></code> </p>
        </td>
        <td><p>Splice junction data.frame summary</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stackJunctions.html">stackJunctions()</a></code> </p>
        </td>
        <td><p>Stack the y-axis position of junctions</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-design-functions" class="hasAnchor"><a href="#section-design-functions" class="anchor"></a>Design functions</h2>
          <p class="section-desc"><p>Functions aimed at manipulation of experimental design</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="curateDFtoDF.html">curateDFtoDF()</a></code> </p>
        </td>
        <td><p>Curate data.frame into a data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="curateVtoDF.html">curateVtoDF()</a></code> </p>
        </td>
        <td><p>Curate vector into a data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="groups2contrasts.html">groups2contrasts()</a></code> </p>
        </td>
        <td><p>Define experimental contrasts from sample groups</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-plot-functions" class="hasAnchor"><a href="#section-plot-functions" class="anchor"></a>Plot functions</h2>
          <p class="section-desc"><p>Functions that assist with R plotting</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="bgaPlotly3d.html">bgaPlotly3d()</a></code> </p>
        </td>
        <td><p>Create an interactive 3-D BGA plotly visualization</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="factor2label.html">factor2label()</a></code> </p>
        </td>
        <td><p>Convert factor to a factor label</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene2gg.html">gene2gg()</a></code> </p>
        </td>
        <td><p>Gene GRangesList to ggplot2 grob</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="grl2df.html">grl2df()</a></code> </p>
        </td>
        <td><p>GRangesList to data.frame for ggplot2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="jitter_norm.html">jitter_norm()</a></code> </p>
        </td>
        <td><p>Apply jitter using normal distribution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSashimi.html">plotSashimi()</a></code> </p>
        </td>
        <td><p>Jam Sashimi plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareSashimi.html">prepareSashimi()</a></code> </p>
        </td>
        <td><p>Prepare Sashimi plot data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stackJunctions.html">stackJunctions()</a></code> </p>
        </td>
        <td><p>Stack the y-axis position of junctions</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-ggplot-functions" class="hasAnchor"><a href="#section-ggplot-functions" class="anchor"></a>ggplot2 functions</h2>
          <p class="section-desc"><p>Functions specific to ggplot2 visualization</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="gene2gg.html">gene2gg()</a></code> </p>
        </td>
        <td><p>Gene GRangesList to ggplot2 grob</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geom_diagonal_wide_arc.html">stat_diagonal_wide_arc()</a></code> <code><a href="geom_diagonal_wide_arc.html">geom_diagonal_wide_arc()</a></code> </p>
        </td>
        <td><p>Draw an area defined by an upper and lower diagonal into an arc</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSashimi.html">plotSashimi()</a></code> </p>
        </td>
        <td><p>Jam Sashimi plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="splicejam-extensions.html">splicejam-extensions</a></code> </p>
        </td>
        <td><p>splicejam extensions to ggforce</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="to_basic.GeomShape.html">to_basic(<i>&lt;GeomShape&gt;</i>)</a></code> </p>
        </td>
        <td><p>Support plotly for GeomShape</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-list-functions" class="hasAnchor"><a href="#section-list-functions" class="anchor"></a>List functions</h2>
          <p class="section-desc"><p>Functions that assist with list manipulation</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="intercalate.html">intercalate()</a></code> </p>
        </td>
        <td><p>Intercalate list values into a vector</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="list2im.html">list2im()</a></code> </p>
        </td>
        <td><p>Convert list to incidence matrix</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-string-functions" class="hasAnchor"><a href="#section-string-functions" class="anchor"></a>String functions</h2>
          <p class="section-desc"><p>Functions aimed at manipulation of character strings</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="escapeWhitespaceRegexp.html">escapeWhitespaceRegexp()</a></code> </p>
        </td>
        <td><p>Escape whitespace in regular expression patterns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sortSamples.html">sortSamples()</a></code> </p>
        </td>
        <td><p>Sort biological sample labels for experimental design</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="strsplitOrdered.html">strsplitOrdered()</a></code> </p>
        </td>
        <td><p>Split the elements of an ordered factor vector</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-spatial-functions" class="hasAnchor"><a href="#section-spatial-functions" class="anchor"></a>Spatial functions</h2>
          <p class="section-desc"><p>Functions that assist with spatial data, particularly 3-D coordinates</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="bgaPlotly3d.html">bgaPlotly3d()</a></code> </p>
        </td>
        <td><p>Create an interactive 3-D BGA plotly visualization</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compressPolygonM.html">compressPolygonM()</a></code> </p>
        </td>
        <td><p>Compress genome coordinates of a matrix of polygons</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dfWide2segments.html">dfWide2segments()</a></code> </p>
        </td>
        <td><p>Convert data.frame to plotly line segment format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simplifyXY.html">simplifyXY()</a></code> </p>
        </td>
        <td><p>Simplify XY coordinates to minimal line segments</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spline3d.html">spline3d()</a></code> </p>
        </td>
        <td><p>Calculate a spline curve fit in 3-D</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-color-functions" class="hasAnchor"><a href="#section-color-functions" class="anchor"></a>Color functions</h2>
          <p class="section-desc"><p>Functions that assist with color manipulation</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="df2colorSub.html">df2colorSub()</a></code> </p>
        </td>
        <td><p>Convert data.frame to categorical colors</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-matrix-functions" class="hasAnchor"><a href="#section-matrix-functions" class="anchor"></a>Matrix functions</h2>
          <p class="section-desc"><p>Functions aimed at manipulation of matrix data</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="shrinkMatrix.html">shrinkMatrix()</a></code> </p>
        </td>
        <td><p>Shrink numeric matrix by groups of rows</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-numeric-functions" class="hasAnchor"><a href="#section-numeric-functions" class="anchor"></a>Numeric functions</h2>
          <p class="section-desc"><p>Functions aimed at manipulation of numeric data</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="geomean.html">geomean()</a></code> </p>
        </td>
        <td><p>Classical geometric mean</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="jamGeomean.html">jamGeomean()</a></code> </p>
        </td>
        <td><p>Modified geometric mean for positive and negative values</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-splicejam-test-data" class="hasAnchor"><a href="#section-splicejam-test-data" class="anchor"></a>splicejam test data</h2>
          <p class="section-desc"><p>R data with exons, coverage, and splice junctions</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="test_cov_gr.html">test_cov_gr</a></code> </p>
        </td>
        <td><p>Sample exon coverage data GRanges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_cov_wide_gr.html">test_cov_wide_gr</a></code> </p>
        </td>
        <td><p>Sample exon coverage data GRanges with wide introns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_exon_gr.html">test_exon_gr</a></code> </p>
        </td>
        <td><p>Sample exon data GRanges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_exon_wide_gr.html">test_exon_wide_gr</a></code> </p>
        </td>
        <td><p>Sample exon data GRanges with wide introns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_junc_gr.html">test_junc_gr</a></code> </p>
        </td>
        <td><p>Sample junction data GRangesList</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_junc_wide_gr.html">test_junc_wide_gr</a></code> </p>
        </td>
        <td><p>Sample junction data GRangesList with wide introns</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-import-functions" class="hasAnchor"><a href="#section-import-functions" class="anchor"></a>Import functions</h2>
          <p class="section-desc"><p>Functions to import specific data formats</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="import_juncs_from_bed.html">import_juncs_from_bed()</a></code> </p>
        </td>
        <td><p>Import splice junction data from BED or SJ.out.tab file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="psl2df.html">psl2df()</a></code> </p>
        </td>
        <td><p>Convert PSL alignment to data.frame</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward, Shannon Farris.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


