<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make tx2gene data.frame from a GTF file — makeTx2geneFromGtf • splicejam</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Make tx2gene data.frame from a GTF file — makeTx2geneFromGtf" />
<meta property="og:description" content="Make tx2gene data.frame from a GTF file" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.58.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make tx2gene data.frame from a GTF file</h1>
    <small class="dont-index">Source: <a href='https://github.com/jmw86069/splicejam/blob/master/R/jambio-tx.R'><code>R/jambio-tx.R</code></a></small>
    <div class="hidden name"><code>makeTx2geneFromGtf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Make tx2gene data.frame from a GTF file</p>
    </div>

    <pre class="usage"><span class='fu'>makeTx2geneFromGtf</span>(
  <span class='no'>GTF</span>,
  <span class='kw'>geneAttrNames</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene_id"</span>, <span class='st'>"gene_name"</span>, <span class='st'>"gene_type"</span>),
  <span class='kw'>txAttrNames</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"transcript_id"</span>, <span class='st'>"transcript_type"</span>),
  <span class='kw'>geneFeatureType</span> <span class='kw'>=</span> <span class='st'>"gene"</span>,
  <span class='kw'>txFeatureType</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"transcript"</span>, <span class='st'>"mRNA"</span>),
  <span class='kw'>nrows</span> <span class='kw'>=</span> -<span class='fl'>1L</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>GTF</th>
      <td><p>character file name sent to <code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>data.table::fread()</a></code>. When the
file ends with ".gz", the <code>R.utils</code> package is recommended, otherwise
the fallback option is to make a system call to <code>gzcat</code>
to gunzip the file during the import step. Note this process fails
when <code>gzcat</code> is not available in the path of the user environment.
In general, the <code>R.utils</code> package is the best solution.</p></td>
    </tr>
    <tr>
      <th>geneAttrNames</th>
      <td><p>character vector of recognized attribute names
as they appear in column 9 of the GTF file, for gene rows.</p></td>
    </tr>
    <tr>
      <th>txAttrNames</th>
      <td><p>character vector of recognized attribute names
as they appear in column 9 of the GTF file, for transcript rows.</p></td>
    </tr>
    <tr>
      <th>geneFeatureType</th>
      <td><p>character value to match column 3 of the GTF
file, used to define gene rows, by default "gene".</p></td>
    </tr>
    <tr>
      <th>txFeatureType</th>
      <td><p>character value to match column 3 of the GTF
file, used to define gene rows, by default "transcript". In some
GTF files, "mRNA" is used, so either is accepted by default.</p></td>
    </tr>
    <tr>
      <th>nrows</th>
      <td><p>integer number of rows to read from the GTF file, by default
-1 means all rows are imported. This parameter is useful to check the
results of a large GTF file using only a subset portion of the file.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical whether to print verbose output during processing.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>data.frame</code> with colnames indicated by the values in
<code>geneAttrNames</code> and <code>txAttrNames</code>.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Create a transcript-to-gene data.frame from a GTF file, which is required
by a number of transcriptome analysis methods such as those in
the DEXseq package, and the limma package functions such as
<code>diffSplice()</code>.</p>
<p>This function also only uses <code><a href='https://Rdatatable.gitlab.io/data.table/reference/fread.html'>data.table::fread()</a></code> and does not
import the full GTF file using something like Bioconductor
<code>GenomicFeatures</code>, simply because the data.table method is markedly
faster when importing only the transcript-to-gene relationship. Also, this
method allows the import of more annotations than are supported by the
typical Bioconductor <code><a href='https://rdrr.io/pkg/rtracklayer/man/export.html'>rtracklayer::import()</a></code> for GTF data.</p>
<p>This function is intended to help keep all transcript data consistent by
using the same GTF file that is also used by other analysis tools, whether
those tools be based in R or more likely, outside R in a terminal
environment.</p>
<p>For example, the GTF file could be used:</p>
<ul>
<li><p>to run STAR sequence alignment
then <code>Rsubread::featureCounts()</code> to generate a matrix of read
counts per gene, transcript, or exon; or</p></li>
<li><p>to generate a transcript
FASTA sequence file then run a kmer quantitation tool such as
Salmon or Kallisto, then using <code>tximport::tximport()</code> to import
results into R for downstream processing.</p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other jam RNA-seq functions: 
<code><a href='assignGRLexonNames.html'>assignGRLexonNames</a>()</code>,
<code><a href='closestExonToJunctions.html'>closestExonToJunctions</a>()</code>,
<code><a href='combineGRcoverage.html'>combineGRcoverage</a>()</code>,
<code><a href='defineDetectedTx.html'>defineDetectedTx</a>()</code>,
<code><a href='detectedTxInfo.html'>detectedTxInfo</a>()</code>,
<code><a href='exoncov2polygon.html'>exoncov2polygon</a>()</code>,
<code><a href='flattenExonsBy.html'>flattenExonsBy</a>()</code>,
<code><a href='groups2contrasts.html'>groups2contrasts</a>()</code>,
<code><a href='make_ref2compressed.html'>make_ref2compressed</a>()</code>,
<code><a href='runDiffSplice.html'>runDiffSplice</a>()</code>,
<code><a href='sortSamples.html'>sortSamples</a>()</code>,
<code><a href='spliceGR2junctionDF.html'>spliceGR2junctionDF</a>()</code></p></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward, Shannon Farris.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


